packages/relay/src/lib/services/cacheService/cacheService.ts

```diff
@@ -0,0 +1,187 @@
+   *
+   * @private
+   */
+  private readonly internalCache: ICacheClient;
+
+  /**
+   * The Redis cache used for caching items from requests.
+   *
+   * @private
+   */
+  private readonly sharedCache: ICacheClient;
+
+  /**
+   * The logger used for logging all output from this class.
+   * @private
+   */
+  private readonly logger: Logger;
+
+  /**
+   * The metrics register used for metrics tracking.
+   * @private
+   */
+  private readonly register: Registry;
+
+  /**
+   * Used for reference to the state of REDIS_ENABLED and REDIS_URL env. variables.
+   */
+  private readonly isSharedCacheEnabled: boolean;
+  /**
+   * Represents a caching manager that utilizes various cache implementations based on configuration.
+   * @param {Logger} logger - The logger used for logging all output from this class.
+   * @param {Registry} register - The metrics register used for metrics tracking.
+   */
+  public constructor(logger: Logger, register: Registry) {
+    this.logger = logger;
+    this.register = register;
+
+    this.internalCache = new LocalLRUCache(logger.child({ name: 'localLRUCache' }), register);
+    this.sharedCache = this.internalCache;
+    this.isSharedCacheEnabled = this.isRedisEnabled();
+
+    if (this.isSharedCacheEnabled) {
+      this.sharedCache = new RedisCache(logger.child({ name: 'redisCache' }), register);
+    }
+  }
+
+  /**
+   * Checks whether Redis caching is enabled based on environment variables.
+   * @private
+   * @returns {boolean} Returns true if Redis caching is enabled, otherwise false.
+   */
+  private isRedisEnabled(): boolean {
+    const redisEnabled = process.env.REDIS_ENABLED && process.env.REDIS_ENABLED === 'true';
+    const redisUrlValid = process.env.REDIS_URL && process.env.REDIS_URL !== '';
+
+    if (redisEnabled && redisUrlValid) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Retrieves a cached value associated with the given key.
+   * If the shared cache is enabled and an error occurs while accessing it,
+   * the internal cache is used as a fallback.
+   * @param {string} key - The key associated with the cached value.
+   * @param {string} callingMethod - The name of the method calling the cache.
+   * @param {string} requestIdPrefix - A prefix to include in log messages (optional).
+   * @param {boolean} shared - Whether to use the shared cache (optional, default: false).
+   * @returns {*} The cached value if found, otherwise null.
+   */
+  public async get(key: string, callingMethod: string, requestIdPrefix?: string, shared: boolean = false): Promise<any> {
+    if (shared && this.isSharedCacheEnabled) {
+      try {
+        return await this.sharedCache.get(key, callingMethod, requestIdPrefix);
+      } catch (error) {
+        const redisError = new RedisCacheError(error);
+        this.logger.error(`Error occured while getting the cache from Redis. Fallback to internal cache. Error is: ${redisError.fullError}`);
+      }
```

Comment:
The logic in this class slightly differs from the agreed design.
As shown here the current conception is "if it's not in the shared then check the local". 
That won't always hold, items should be allowed to exist in local only without first checking in shared, items in shared shouldn't all have to be in local.
Services should be able to use both caches as they see fit. As coded a service cannot dictate it's own population and consideration for local and shared.

You could expose a public getter for the 2 caches and allow services to call what they need e.g. `await this.cacheService.local().get(...)` or `await this.cacheService.shared().get(...)`

Reply:
If you want their to be 1 default then you can based on `this.isSharedCacheEnabled` but you'd still want to expose the 2 and only interact with 1 at a time.

Reply:
Not sure i understand, but the logic is the same in the design doc -> "by default always go to internal cache, except if the method didn't explicitly asked for shared, by setting the shared variable to true" : 
1. based on shared variable, which is set to true by the method and if it wants to address the Redis or internal. 
2. If shared is true, go to shared cache. If redis is not available for some reason or it returns an error, then fallback to internal
3. If shared is false, go to internal

This way method can decide what cache it wants to use and cacheService takes care of using it and error handling.
By exposing getters and using them by methods (in filter service or subscribe service or somewhere else) the whole cache Service is not needed, because it's main reason is to hide implementation from external services (that way if we change something, we don't need to change it everywhere as long as we keep following the interface) and another main reason is the error handling, if something goes wrong with the shared cache to fallback and report.

Reply:
By exposing getters and using them by methods (in filter service or subscribe service or somewhere else) the whole cache service is not needed
Yes this is one of the items I'm trying to resolve with my suggestion.
I missed the `shared` param so i'm re-reviewing this and thinking through how it scales

Reply:
Resolving this as I'm on board now that i've re-reviewed